app/src/main/java/com/nervousfish/nervousfish/activities/QRExchangeKeyActivity.java

```diff
@@ -0,0 +1,251 @@
+     *
+     * @param savedInstanceState Don't touch this
+     */
+    @Override
+    protected void onCreate(final Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        this.setContentView(R.layout.activity_qrexchange);
+
+        final Intent intent = getIntent();
+        this.serviceLocator = (IServiceLocator) intent.getSerializableExtra(ConstantKeywords.SERVICE_LOCATOR);
+
+
+        //TODO: Get the user's generated public key from the database
+        final IKeyGenerator keyGenerator = serviceLocator.getKeyGenerator();
+        final KeyPair pair = keyGenerator.generateRSAKeyPair("test");
+        publicKey = pair.getPublicKey();
+
+
+        final Button scanButton = (Button) findViewById(R.id.scanbutton);
+        scanButton.setOnClickListener(new View.OnClickListener() {
+
+            @Override
+            public void onClick(final View v) {
+                LOGGER.info("Started scanning QR code");
+                final IntentIntegrator integrator = new IntentIntegrator(QRExchangeKeyActivity.this);
+                integrator.initiateScan();
+
+            }
+        });
+
+        final Button generateButton = (Button) findViewById(R.id.generateQRbutton);
+        generateButton.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(final View v) {
+                LOGGER.info("Started generating QR code");
+                final String space = " ";
+                final Bitmap qrCode = QRGenerator.encode(publicKey.getType() + space + publicKey.getName()
+                        + space + publicKey.getKey());
+                showQRCode(qrCode);
+            }
+        });
+
+        final ImageButton backButton = (ImageButton) findViewById(R.id.backButtonQRExchange);
+        backButton.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(final View v) {
+                finish();
+            }
+        });
+
+    }
+
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void onActivityResult(final int requestCode, final int resultCode, final Intent intent) {
+        final IntentResult scanResult = IntentIntegrator.parseActivityResult(requestCode, resultCode, intent);
+        if (scanResult != null) {
+            final String result = scanResult.getContents();
+            addNewContact(result);
+
+        } else {
+            LOGGER.error("No scan result in QR Scanner");
+        }
+
+
+    }
+
+
+    /**
+     * Adds new contact with the scanned key and opens change
+     *
+     * @param qrMessage The information we got from the QR code.
+     */
+    private void addNewContact(final String qrMessage) {
+        //TODO: Add recognizer for contact name to avoid saving the same key twice (add your personal name to QR code)
+        final IKey key = QRGenerator.deconstructToKey(qrMessage);
+        final EditText editName = new EditText(this);
```

Comment:
I think it's better to move this code to a separate helper class, because NFCActivity and BluetoothActivity should have literally exactly the same code